pkg/ipam/clusterpool.go

```diff
@@ -0,0 +1,716 @@
+			}
+		}
+	} else {
+		// Iterate over pod CIDRs in reverse order so we prioritize releasing
+		// later pod CIDRs.
+		for i := len(p.ipAllocators) - 1; i >= 0; i-- {
+			ipAllocator := p.ipAllocators[i]
+			cidrNet := ipAllocator.CIDR()
+			cidrStr := cidrNet.String()
+			if _, released := p.released[cidrStr]; released {
+				continue
+			}
+			var status types.UsedPodCIDRStatus
+			if i == 0 || ipAllocator.Used() > 0 {
+				// If this is the first pod CIDR or it is used, then mark it as
+				// in-use or depleted.
+				if ipAllocator.Free() == 0 {
+					status = types.PodCIDRStatusDepleted
+				} else {
+					status = types.PodCIDRStatusInUse
+				}
+			} else if _, removed := p.removed[cidrStr]; removed {
+				// Otherwise, if the pod CIDR has been removed, then mark it as released.
+				p.released[cidrStr] = struct{}{}
+				delete(p.removed, cidrStr)
+				status = types.PodCIDRStatusReleased
+				log.WithField(logfields.CIDR, cidrStr).Debug("releasing removed pod CIDR")
+			} else if free := ipAllocator.Free(); totalFree-free >= p.releaseThreshold {
+				// Otherwise, if the pod CIDR is not used and releasing it would
+				// not take us below the release threshold, then release it and
+				// mark it as released.
+				p.released[cidrStr] = struct{}{}
+				totalFree -= free
+				status = types.PodCIDRStatusReleased
+				log.WithField(logfields.CIDR, cidrStr).Debug("releasing pod CIDR")
+			} else {
+				// Otherwise, mark the pod CIDR as in-use.
+				status = types.PodCIDRStatusInUse
+			}
+			result[cidrStr] = types.UsedPodCIDR{
+				Status: status,
+			}
+		}
+	}
+
+	return result
+}
+
+func (p *podCIDRPool) updatePool(podCIDRs []string) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if option.Config.Debug {
+		log.WithFields(logrus.Fields{
+			logfields.NewCIDR: podCIDRs,
+			logfields.OldCIDR: p.inUsePodCIDRsLocked(),
+		}).Debug("Updating IPAM pool")
+	}
+
+	// Parse the pod CIDRs, ignoring invalid CIDRs, and de-duplicating them.
+	cidrNets := make([]*net.IPNet, 0, len(podCIDRs))
+	cidrStrSet := make(map[string]struct{}, len(podCIDRs))
+	for _, podCIDR := range podCIDRs {
+		_, cidr, err := net.ParseCIDR(podCIDR)
+		if err != nil {
+			log.WithError(err).WithField(logfields.CIDR, podCIDR).Error("ignoring invalid pod CIDR")
+			continue
+		}
+		if _, ok := cidrStrSet[cidr.String()]; ok {
+			log.WithField(logfields.CIDR, podCIDR).Error("ignoring duplicate pod CIDR")
+			continue
+		}
+		cidrNets = append(cidrNets, cidr)
+		cidrStrSet[cidr.String()] = struct{}{}
+	}
+
+	// Forget any released pod CIDRs no longer present in the CRD.
+	for cidrStr := range p.released {
+		if _, ok := cidrStrSet[cidrStr]; !ok {
+			log.WithField(logfields.CIDR, cidrStr).Debug("removing released pod CIDR")
+			delete(p.released, cidrStr)
+		}
+	}
+
+	// Special case: the first call to updatePool() with at least one valid pod
+	// CIDR will create the initial allocator. This first pod CIDR is treated as
+	// a special case and will never be released.
```

Comment:
Yes, you are absolutely right. You reviewed it a second too soon. I have already removed it locally, but haven't pushed yet because I was waiting on CI to finish. Will push an updated version soon.